srv/salt/_modules/osd.py

```diff
@@ -382,28 +392,104 @@ def wait(self):
         i = 0
         last_pgs = 0
         while i < self.settings['timeout']/self.settings['delay']:
+            rc, msg = self.osd_safe_to_destroy()
+            if rc == 0:
+                log.info("osd.{} is safe to destroy".format(self.osd_id))
+                return ""
             entry = self.osd_df()
             if 'pgs' in entry:
                 if entry['pgs'] == 0:
-                    log.info("osd.{} has no PGs".format(self.osd_id))
-                    return ""
+                    log.warning("osd.{} has {} PGs remaining but {}".
+                                format(self.osd_id, entry['pgs'], msg))
                 else:
                     log.warning("osd.{} has {} PGs remaining".format(self.osd_id, entry['pgs']))
                     if last_pgs != entry['pgs']:
                         # Making progress, reset countdown
                         i = 0
                         last_pgs = entry['pgs']
             else:
-                msg = "osd.{} does not exist".format(self.osd_id)
+                msg = "osd.{} does not exist {}".format(self.osd_id, msg)
                 log.warning(msg)
-                return msg
             i += 1
             time.sleep(self.settings['delay'])
 
         log.debug("Timeout expired")
         raise RuntimeError("Timeout expired")
 
 
+class CephPGs(object):
+    """
+    Query PG states and pause until all are active+clean
+    """
+
+    def __init__(self, **kwargs):
+        """
+        Initialize settings, connect to Ceph cluster
+        """
+        self.settings = {
+            'conf': "/etc/ceph/ceph.conf",
+            'timeout': 120,
+            'keyring': '/etc/ceph/ceph.client.admin.keyring',
+            'client': 'client.admin',
+            'delay': 12
+        }
+        self.settings.update(kwargs)
+        log.debug("settings: {}".format(pprint.pformat(self.settings)))
+        self.cluster = rados.Rados(conffile=self.settings['conf'],
+                                   conf=dict(keyring=self.settings['keyring']),
+                                   name=self.settings['client'])
+        try:
+            self.cluster.connect()
+        except Exception as error:
+            raise RuntimeError("connection error: {}".format(error))
+
+    def quiescent(self):
+        """
+        Wait until PGs are active+clean or timeout is reached.  Default is a
+        2 minute sliding window.
+        """
+        i = 0
+        last = []
+        while i < self.settings['timeout']/self.settings['delay']:
```

Comment:
Only because we use it in a division here and its a config, maybe check that delay is not 0? Pretty sure some of our usual suspects will try to set delay to 0 at some point...to make things go faster.

Reply:
I thought of that for a second and wondered exactly what somebody would expect the correct behavior is?  That we would try to busy wait a cpu?  

Do you think this is just deserving of a specific exception?  That is, RuntimeError("Delay cannot be 0")?

Reply:
I think the built-in ValueError would be appropriate.